Integrating CKEditor

AllanYangZhou opened the pull request:
Integrating CKEditor as our new RTE. Primary advantage over existing RTE is the component previews.

Directly addresses: #1810, #1811, #1933
Also addresses: #1535, #1431

![image](https://cloud.githubusercontent.com/assets/8732147/15629986/97c55a90-24dd-11e6-8861-f691a2f237fa.png)

seanlip commented:
Hi @AllanYangZhou, thanks! Just to quickly check: what do you want me to look out for / comment on, in this review?

Also, /cc @kahkhang.

AllanYangZhou commented:
Primary thing I wasn't sure about was how I was integrating in the new code, especially the custom widget stuff. If there's any files/code that are out of place and make more sense somewhere else, please let me know.

Also if you could, please take a look at the system for actually displaying components in `plugin.js`. Currently I essentially handcrafting an element to mimic the component's angular directive in the editor. This isn't a big deal with links, but it might get painful for some of the other ones. Maybe there is a smarter way to do this? Or to get angular directives to actually show up?

seanlip commented on extensions/rich_text_components/Link/static/oppialink/plugin.js:
```diff
@@ -0,0 +1,53 @@
+    editor.widgets.add('oppialink', {
+      button: 'Add a link',
+      inline: true,
+      template: '<oppia-noninteractive-link ' +
```
I wonder if the computation of this template could be parameterized somehow based on the RTE components dict?

seanlip replied:
(You could pass in a dummy RTE components dict, declared as an oppia.constant, to test this.)

seanlip commented on extensions/rich_text_components/Link/static/oppialink/plugin.js:
```diff
@@ -0,0 +1,53 @@
+        el.setHtml('');
+      },
+      data: function() {
+        this.element.setAttribute(
```
Ditto, ideally most of this will be autogenerated from a rte-components-dict.

seanlip commented on extensions/rich_text_components/Link/static/oppialink/plugin.js:
```diff
@@ -0,0 +1,53 @@
+          'url-with-value', `&quot;${this.data.url}&quot;`);
+
+        var display = this.data.text || this.data.url;
+        this.element.setHtml(`<a>${display}</a>`);
```
Does the 'a' here need to be handcrafted? If so, that might be a problem...

seanlip commented on extensions/rich_text_components/Link/static/oppialink/plugin.js:
```diff
@@ -0,0 +1,53 @@
+        this.element.setHtml(`<a>${display}</a>`);
+      },
+      init: function() {
+        var openInSame = this.element.getAttribute(
```
This should be autogenerated from a declarative representation of the attributes (i.e. the rte-components-dict).

seanlip commented:
Thanks, @AllanYangZhou! I took a look -- it looks mostly fine; I think the case would be stronger if some of the stuff were autogenerated as I mentioned, but that looks doable in theory.

The main concern I have is whether the `<a>` tag is necessary. If so, then that seems like it might be an issue, because there's already supposed to be an `<a>` tag in the directive HTML. Is this what you referred to by handcrafting? If so, yes, that's problematic, and it might be worth finding out how, e.g., CKEditor manages to display math expressions (that presumably use the third-party library MathJax). Or maybe we decide to rewrite all the RTE components in plain JavaScript...

AllanYangZhou commented:
@seanlip Thanks for taking a look!

To clarify, the handcrafting is specifically for the purpose of displaying the widget's preview, and is not (necessarily) related to the widget's actual source code, which in this case would be `<oppia-noninteractive-link ...></oppia-noninteractive-link`. In essence, for each widget CKEditor can maintain separately the display in editor and the actual source. So in the worst case, we could always use the image previews that Andrew has created for the previews.

But I agree that needing to handcraft previews for each widget is problematic, and ideally we can figure out a way to directly render the directives as their own preview.

AllanYangZhou commented:
@seanlip and @kahkhang can you take a look at the recent changes?

I believe I have largely solved the issue of rendering the rte component directives as previews in CKEditor. This is great news because we no longer have to handcraft previews for every component. I've also rewritten the widget code so that widgets are automatically generated from the rte components dict.

Every component's preview seems to be functional right now, though there are likely some minor bugs still (I've noticed some console errors from the video component).

seanlip commented:
Hi @AllanYangZhou -- I'm running into some trouble after checking out this branch and installing it. In particular, I ran into an error like this three times in succession when running it using start.sh:

Downloading and unzipping file lineutils_4.5.8 to ./third_party/static
Traceback (most recent call last):
  File "scripts/install_third_party.py", line 331, in <module>
    _install_third_party_libs()
  File "scripts/install_third_party.py", line 319, in _install_third_party_libs
    download_manifest_files(MANIFEST_FILE_PATH)
  File "scripts/install_third_party.py", line 297, in download_manifest_files
    dependency_zip_root_name, dependency_target_root_name)
  File "scripts/install_third_party.py", line 139, in download_and_unzip_files
    os.path.join(target_parent_dir, target_root_name))
OSError: [Errno 2] No such file or directory

Do you encounter this too when you do a fresh install? I'm using Linux Ubuntu.

AllanYangZhou commented:
Yeah, I just removed my third_party/static/ and started up the server and ran into the same problem. I get errors the first 3 times (one for each of the new packages I've added) but not the 4th. Trying to figure out what it is right now.

AllanYangZhou commented:
@seanlip I think I've fixed the install issue, thanks for pointing it out. Let me know if you still run into any issues.

seanlip commented:
Thanks -- it works now! :-)

I played with it a bit, and found a few issues:
- When I press Ctrl-Z, and there's already a video or a math formula in the area, they flicker. Also, Ctrl-Z sometimes works and sometimes doesn't. I don't think it works on rich-text components?
- It's interesting that the previews are playable. I can't seem to edit the parameters for the video though -- when I double click on it, the customization modal doesn't open. Do you think it's worth putting a nonclickable mask over all the previews, in edit mode?
- Is the position of the toolbar configurable? It also seems to go 'over' the customization modal briefly when I double-click on a component, but maybe that can be solved by adjusting the z-indexes.
- When I put some text, an image component, and then a tabs component in the RTE, and use the up/down or left/right arrow keys to move between them, the cursor kind of disappears. I can't seem to get the cursor between the image and tabs, either (in order to add a newline between them) -- it appears that the cursor is sometimes "on" the rich-text component, and hitting enter opens the customization modal. This is a bit unintuitive.

That's all I found, but otherwise it looks promising! Let me know if any of the above doesn't make sense; I'd be happy to elaborate. Thanks!

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +724,199 @@ oppia.config(['$provide', function($provide) {
+          editor.widgets.add(ckName, {
+            button: componentDefn.tooltip,
+            inline: inlineCapable.indexOf(componentDefn.name) !== -1,
+            template: `<${tagName}></${tagName}>`,
```
`...` is not supported at all in IE, I think. Any alternatives?

AllanYangZhou replied:
Good point, changed it.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +724,199 @@ oppia.config(['$provide', function($provide) {
+                '</div>',
+      require: '?ngModel',
+      link: function(scope, el, attr, ngModel) {
+        var widgets = [
```
richTextComponents; please also rename below. I think we should only use 'widgets' when absolutely necessary, i.e. when mandated by CKEditor.

AllanYangZhou replied:
Done.

seanlip commented on core/tests/karma.conf.js:
```diff
@@ -18,6 +18,7 @@ module.exports = function(config) {
       'third_party/static/angularjs-1.4.7/angular.js',
       'third_party/static/angularjs-1.4.7/angular-mocks.js',
       'third_party/static/math-expressions-762ffd/build/math-expressions.js',
+      'third_party/static/ckeditor/ckeditor.js',
```
There ought to be a version number at the end of ckeditor...

AllanYangZhou replied:
Done.

seanlip commented on manifest.json:
```diff
@@ -280,6 +280,27 @@
+        "downloadFormat": "zip",
+        "url": "http://download.ckeditor.com/widget/releases/widget_4.5.8.zip",
+        "rootDir": "widget",
+        "targetDir": "widget"
```
Can this be ckeditor-widget, ckeditor-lineutils? I think it's good to tie this back to ckeditor in order to avoid confusion.

AllanYangZhou replied:
Done.

seanlip commented on manifest.json:
```diff
@@ -280,6 +280,27 @@
+        "rootDir": "ckeditor",
+        "targetDir": "ckeditor"
+      },
+      "widget": {
```
I'm not sure, but is it possible for this to be ckeditorWidget? Ditto below.

AllanYangZhou replied:
Done.

AllanYangZhou commented:
Hi Sean, thanks for taking a look! Wrt to your bullet points:

* **Undo behavior**:
    * For the flickering, this is a result of angular rerendering the components, which currently happens to all the components whenever any of them change. I plan on changing this so only the changed component gets rerendered, but there is no way to completely get rid of the flickering.
    * Yes, the Ctrl-Z behavior seems to be somewhat unpredictable for the rich text components, I'm actually not sure how to approach this
* **Playable previews**: for all the "block" components (video, images, tabs, collapsible) I've added a div overlay to prevent any clicks from going through. This also fixes the issue where you couldn't edit the video component. Though I think allowing interactivity with tabs and collapsible previews might be useful. What do you think?
* **Toolbar positioning**: the position of the toolbar is configurable, I've put it 15px above the main content (see below). Yeah the floating toolbar was a nuisance I think, I've replaced it with a fixed one (see below).
![image](https://cloud.githubusercontent.com/assets/8732147/15447355/93ba992c-1ef4-11e6-8c6f-8021b3f7ec9c.png)
* **Issue moving cursor between two block components**:
    * when there are two "block" components right next to each other, there is no `<p>` in between them so you cannot move your cursor there. The reason why the cursor disappears is that is "on" the components (you have selected it). I've changed it so that the selected component has a blue border, making this more obvious.
    * It is still possible to type in between them: if you hover your mouse between the two components, a red line pops up with a button to "Insert paragraph here" which will insert a line in between the two. I agree this is not intuitive--one thing I can try to do is automatically insert the line in between for them, but this may not be very intuitive either (what if they didn't want the line in between the two?)

seanlip commented:
Thanks, @AllanYangZhou!

- I think flickering for only the changed component is probably fine. However, the Ctrl-Z behavior not working for components is a pretty serious downside...
- Hm, not sure. Maybe for simplicity let's prevent all clicks for now, and relax this later if asked.
- Great!
- I think the default behaviour is probably OK, as long as it is clearly indicated (e.g. what you've done with the blue border). It's not something that happens very often, so I'm not too worried.

So, I think the main issue left is the Ctrl-Z. (Is there also a corresponding "redo" functionality?)

Also, @kahkhang -- do you have any feedback?

Thanks, @AllanYangZhou!

kahkhang commented:
Hi @AllanYangZhou and @seanlip, the CKEditor looks great!

I've noticed some minor issues:
- After a collapsible block and a fraction is inserted to a blank editor, inserting a link throws an error (which can be resolved by inserting the link again, so its pretty minor). I'm thinking it might be related to the rendering of the angular components, since you can trigger the same error by opening a tab component dialog, and inserting a tab component dialog again in one of the editors in the dialog
<img width="1920" alt="screenshot 2016-05-23 20 19 58" src="https://cloud.githubusercontent.com/assets/4048170/15491327/3010dce2-2124-11e6-90ca-ff7acce089d6.png">

- If a component is inserted at the start, the cursor cannot be moved to the start of the editor (It can be resolved by hovering over the component and clicking on the red button that pops up, but I agree it is not very intuitive)
<img width="631" alt="screenshot 2016-05-23 20 29 18" src="https://cloud.githubusercontent.com/assets/4048170/15491416/09a0d3a4-2125-11e6-9b8c-3006c06717ea.png">

- The cursor jumps to the end of the line when i try to click to the left of the link in the setup below, but typing a character puts the cursor back to its intended position
<img width="635" alt="screenshot 2016-05-23 20 33 51" src="https://cloud.githubusercontent.com/assets/4048170/15491487/b05c089e-2125-11e6-8402-c8fa2f33524e.png">

Thanks!

AllanYangZhou commented:
@kahkhang Thanks for taking a look!

* This one is interesting, and I couldn't figure out the cause. Ultimately, I tried circumventing the issue by programmatically moving the caret past the widget once it is inserted, and it appears the issue no longer occurs.
* Yeah, this is similar to the issue discussed earlier with trying to place the cursor between two block components. Ultimately I think the red line is a workable solution for now, we could add some way to do this automatically in the future.
* Yeah, I think we encountered something similar during #1426? I think we did some CSS hack to fix it. Though I think this is minor enough to leave alone for now.

AllanYangZhou commented:
@seanlip PTAL

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2931,3 +2931,21 @@ md-card.preview-conversation-skin-supplemental-card {
   margin-bottom: 10px;
 }
+
+.component-container {
```
Are these styles from ckeditor or are they ours?

If ours, add .oppia-rte prefix; if ckeditor's, add comment explaining this.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2931,3 +2931,21 @@ md-card.preview-conversation-skin-supplemental-card {
+}
+
+.component-overlay {
+  position: absolute;
```
please alphabetize attrs within each selector

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2931,3 +2931,21 @@ md-card.preview-conversation-skin-supplemental-card {
+  height: 100%;
+}
+
+.ckeditor-rte{
```
nit: add space before {

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -2931,3 +2931,21 @@ md-card.preview-conversation-skin-supplemental-card {
+}
+
+.ckeditor-rte{
+    outline: 1px solid gray;
```
nits: alphabetize; indentation is wrong

AllanYangZhou replied:
Fixed.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -665,34 +668,12 @@ oppia.factory('rteHelperService', [
-      var _openCustomizationModal = function(
+      },
+      isInlineComponent: function(richTextComponent) {
+        var inlineComponents = ['link', 'math'];
```
Declare at top, and add to it in the Object.keys(RTE_COMPONENT_SPECS).forEach() call? Seems like this is something that ought to be declared in the component config...

AllanYangZhou replied:
Added `is_inline` to the component definition in backend.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+            button: componentDefn.tooltip,
+            inline: isInline,
+            template: componentTemplate,
+            edit: function(event) {
```
We tend to use 'evt' instead, since 'event' is an IE global var.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+                    }
+                  }
+
+                  if (!that.isReady()) {
```
Well... this is a bit confusing. When exactly is creation supposed to happen? In what circumstances do we need the guard here?

AllanYangZhou replied:
`isReady()` checks whether the widget has been inited and the DOM has been set up (see [here](http://docs.ckeditor.com/#!/api/CKEDITOR.plugins.widget-property-ready)). `finalizeCreation` actually inserts the widget in. So when you first click on a button to insert a widget, enter all the appropriate data into the modal, and click submit, then `isReady()` is `false` and we call `finalizeCreation`. If you then double click on the widget to edit it later, `isReady()` is `true` (the widget already exists, we just need to modify some attributes) so we do not need to `finalizeCreation` again.

AllanYangZhou replied:
I've added a comment to clear this up.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+                    editor.widgets.finalizeCreation(container);
+                  }
+                  // Need to manually $compile so the directive renders.
+                  $compile($(that.element.$).contents())($rootScope);
```
What is that.element.$? Just wondering; I've never seen that construct before.

AllanYangZhou replied:
`that.element` in this case is CKEditor's custom representation of a DOM element. The `$` property is the native DOM element it represents. See here [CKEDITOR.dom.element](http://docs.ckeditor.com/#!/api/CKEDITOR.dom.element)

seanlip replied:
Gosh, that's confusing -- everyone seems to use $ for different things.

Maybe add the link you sent me, as a reference?

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+                  // Need to manually $compile so the directive renders.
+                  $compile($(that.element.$).contents())($rootScope);
+                  $timeout(function() {
+                    if (isInline) {
```
Maybe add comment explaining why these two cases are being treated differently?

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+              return element.children[0];
+            },
+            upcast: function(element) {
+              // This is how the widget is recognized by CKEditor.
```
Maybe give more detail about downcast/upcast -- what do their return values signify? I tried searching for "CKEditor downcast" and couldn't easily find documentation for it that didn't explain downcasting in terms of downcasting... (!)

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+      scope: {
+        uiConfig: '&'
+      },
+      template: '<div><div></div>' +
```
Maybe add comment to explain why the extra divs are there?

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+      template: '<div><div></div>' +
+                '<div contenteditable="true" class="ckeditor-rte">' +
+                '</div></div>',
+      require: '?ngModel',
```
why the ?

AllanYangZhou replied:
Yeah it's not necessary I think. I've removed it.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+      require: '?ngModel',
+      link: function(scope, el, attr, ngModel) {
+        var _RICH_TEXT_COMPONENTS = rteHelperService.getRichTextComponents();
+        var names = [];
```
componentNames?

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+    var _RICH_TEXT_COMPONENTS = rteHelperService.getRichTextComponents();
+    _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+      // The name of the CKEditor widget corresponding to this component.
+      var ckName = 'oppia' + componentDefn.name;
```
I think a similar construction is done elsewhere, in line 903. This sets up the potential for skew. May want to extract to a common method, or (perhaps better) make them a backend property.

AllanYangZhou replied:
Added it to the backend component defn.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+          }
+        });
+
+        // Rules to prevent CKEDITOR from filtering components and wrappers.
```
Add link to webpage which explains the rules syntax

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+        }).join(',');
+        var buttonNames = [];
+        names.forEach(function(name) {
+          buttonNames.push('Oppia' + name);
```
Bit of an odd button name? Don't you want e.g. the human-readable description, or is this meant to be an internal property of sorts?

AllanYangZhou replied:
Yeah `buttonNames` is used internally to refer to the buttons to add to the toolbar. Not what users see.

seanlip replied:
So, what does buttonNames actually get used for? (I.e., if I use some random name, what will break?)

AllanYangZhou replied:
If you don't specify all the button names, the component buttons will not appear on the toolbar (see line 953). So, adding a plugin/widget to the editor and adding the button for it are two separate steps, `pluginNames` handles the first step and `buttonNames` handles the second.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+          if (html === undefined) {
+            return html;
+          }
+          return html.replace(componentRe, function(match, p1, p2, p3) {
```
I think p1, p2, p3 could use a little explanation for coders less familiar with regexes.

seanlip replied:
(Just to clarify, by this, I mean a comment explaining what each arg represents conceptually.)

AllanYangZhou replied:
Good idea. I've decided to do this next to where I declare the regex, to avoid repetition.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+
+        // Angular rendering of components confuses CKEditor's undo system, so
+        // we hide all of that stuff away from CKEditor.
+        ck.on('getSnapshot', function(event) {
```
evt

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+          ck.setData(wrapComponents(ngModel.$viewValue));
+        };
+
+        scope.$on('$destroy', function() {
```
+1 :)

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+          if (event.data === undefined) {
+            return;
+          }
+          event.data = event.data.replace(componentRe, function(match, p1, p2) {
```
Just so I understand, where's the corresponding logic that reinstates the full tags when a snapshot is subsequently reinstated?

AllanYangZhou replied:
There is none, all snapshots are filtered of any angular rendering content. When someone hits undo, all the widgets are upcasted (recognized as widgets) and $compile'd.

seanlip replied:
Got it. This seems like useful context to add in a comment, I think.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -665,34 +668,12 @@ oppia.factory('rteHelperService', [
+        var inlineComponents = ['link', 'math'];
+        return inlineComponents.indexOf(richTextComponent) !== -1;
+      },
+      _openCustomizationModal: function(
```
Seems like this is used elsewhere, so probably needn't be private.

AllanYangZhou replied:
Ah yes, fixed.

seanlip commented on core/templates/dev/head/header_js_libs.html:
```diff
@@ -23,3 +23,14 @@
+<script>
+  // Required for custom widgets.
+  CKEDITOR.plugins.addExternal(
+  'widget', '/third_party/static/ckeditor-widget/', 'plugin.js');
```
indent this line by 2. Ditto below.

AllanYangZhou replied:
Done.

seanlip commented:
@souravbadami, @arun1595 -- as creators, can you PTAL at the UI/functionality for this, and let us know what you think? Thanks!

AllanYangZhou commented:
@seanlip I've addressed the comments, PTAL.

Note: there is some focus issue when using CKEditor in a modal (e.g., suggestions) that prevents you from typing. The most common [solution/hack](https://gist.github.com/Reinmar/b9df3f30a05786511a42#ckeditor-inside-bootstrap-modals) involves monkey-patching the modal library itself.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +726,383 @@ oppia.config(['$provide', function($provide) {
+          }
+          return html.replace(componentRe, function(match, p1, p2, p3) {
+            if (rteHelperService.isInlineComponent(p3)) {
+              return '<span type="oppia-noninteractive-' + p3 + '">' +
```
This looks similar to the componentTemplate stuff on lines 758 and 762. Should they be consolidated in order to avoid skew?

seanlip commented on core/templates/dev/head/forms/form_builder_templates.html:
```diff
@@ -219,9 +219,9 @@
 <script type="text/ng-template" id="schemaBasedEditor/html">
-  <text-angular-rte ng-if="!isDisabled()" html-content="$parent.localValue"
+  <ck-editor-rte ng-if="!isDisabled()" ng-model="$parent.localValue"
                     ui-config="uiConfig()">
```
nit: deindent three spaces to align with attr in previous line

seanlip commented on extensions/rich_text_components/base.py:
```diff
@@ -36,6 +36,9 @@ class BaseRichTextComponent(object):
     # Overridden in subclasses.
     category = ''
+    # The component's name in the CKEditor widget system.
+    # Overriden in subclasses.
```
Just to check: if we ever change this, will it affect content that's already in the datastore?

AllanYangZhou replied:
No, it shouldn't.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +726,383 @@ oppia.config(['$provide', function($provide) {
+             * when we first load data in. Returns a boolean,
+             * true iff "element" is an instance of this widget.
+             */
+            upcast: function(element) {
```
Weird naming! So downcast() converts element --> DOM node, and upcast() converts element --> boolean?

I mean, I guess it's what it is, but just wanted to double-check.

AllanYangZhou replied:
`downcast` takes an element and returns another element. You could choose to not do anything at all (just return the element you're given). What it allows is for your widget during editing to be completely different from the widget's representation in the output source. In our usage, block widgets like the video/Youtube component are wrapped in a div and accompanied by an additional overlay div. Thanks to downcasting, neither of these appear in the outputs source.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +726,383 @@ oppia.config(['$provide', function($provide) {
+                dontUpdate: true
+              });
+              var that = this;
+              var isMissingAttributes = false;
```
Suggest 'hasMissingAttributes'

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +726,383 @@ oppia.config(['$provide', function($provide) {
+                      element.children.length > 0 &&
+                      element.children[0].name === tagName);
+            },
+            data: function() {
```
When is this called? Suggest adding docstring.

(It feels slightly weird to me that the method name has no verb and that it returns nothing -- but again, perhaps that's just how it is.)

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -747,9 +726,383 @@ oppia.config(['$provide', function($provide) {
+          skin: 'bootstrapck,/third_party/static/ckeditor-bootstrapck/',
+          toolbar: [
+            {
+              name: 'history',
```
Just to check, are these names arbitrary? If so, might be worth mentioning this.

seanlip commented:
Hi @AllanYangZhou, I think this looks great, thanks! I have mainly trivial comments.

Also noting that, per our discussion, we should sort out the structural format of existing data before merging this in.

jacobdavis11 commented:
Design doc: https://drive.google.com/open?id=1qjWIZ_843K6PL74l6NmpkbQakm3HcvKTA9tG2o-Zt5Q

AllanYangZhou commented:
@seanlip Oops, I meant to bring branch 'ckeditor' up to date with branch 'develop' but when pushing to branch 'ckeditor' I accidentally pushed these PR changes as well, which is why this says it's merged.

I'll be making my formatting changes in a separate PR, and of course there will be a separate final PR to merge branch 'ckeditor' into 'develop'

seanlip commented:
OK, thanks for the note! Hopefully the eventual merge to develop doesn't include so many commits...

AllanYangZhou commented:
@seanlip Should be fine, the reason why there's so many commits here is because 'ckeditor' was so far behind 'develop'
